test/test_dataloader.py

```diff
@@ -175,21 +225,65 @@ def __len__(self):
 
 
 def _test_timeout():
-    os.close(sys.stderr.fileno())
-    sys.stderr.close()
     dataset = SleepDataset(10, 10)
     dataloader = DataLoader(dataset, batch_size=2, num_workers=2, timeout=1)
     _ = next(iter(dataloader))
 
 
 def _test_segfault():
-    os.close(sys.stderr.fileno())
-    sys.stderr.close()
     dataset = SegfaultDataset(10)
     dataloader = DataLoader(dataset, batch_size=2, num_workers=2)
     _ = next(iter(dataloader))
 
 
+def _test_interrupt_retry(timeout=0):
+    dataset = TensorDataset(torch.randn(1, 1), torch.randn(1, 1))
+    dataloader = DataLoader(dataset, batch_size=1, num_workers=1, timeout=timeout)
+    dataloaderiter = iter(dataloader)
+
+    # make SIGUSR1 interrupt
+    def handler(signum, frame):
+        pass
+    signal.signal(signal.SIGUSR1, handler)
+
+    # Replace iterator getter with a wrapper that reliably calls an
+    # interruptable blocking recv syscall to simulate interruption during recv
+    # in queue.get.
+    # The used socket.recv call below in the replacing function is quite
+    # dangerous because it blocks everything on Python side, including the
+    # cleaning up in dataloder.__del__ when this process exits. To prevent
+    # orphan worker child, we manually terminate worker process here.
```

Comment:
Why does waiting on a socket prevents Python from cleaning up after itself? That is unclear to me. Is it because the test will terminate this process before it has a chance to clean up?

Reply:
If I don't terminate the workers here, the test leaves orphan worker processes even if they have`.daemon=True` and we terminate the loader process. Even adding a python SIGTERM handler doesn't work. So my guess is that `socket.recv` holds on GIL or something that stops everything in Python from running.